### PR TITLE
WICKET-5827 - CssUrlReplacer supports base64 encoded images - 6.x

### PR DESCRIPTION
**Pull request by klopfdreh**
http://apache-wicket.1842946.n4.nabble.com/CssUrlReplacer-improvement-base64-content-td4669546.html

### CODE REVIEW

**Comment by klopfdreh** [comment]
Forgot to close the resource stream. In the  method private CharSequence createBase64EncodedImage(PackageResourceReference imageReference).

May wrap all but the first line into try / finally and close the stream. Sorry for that.

File: wicket-core/src/main/java/org/apache/wicket/resource/CssUrlReplacer.java [hunk 1]
```diff
@@ -43,6 +49,8 @@
 	// The pattern to find URLs in CSS resources
 	private static final Pattern URL_PATTERN = Pattern.compile("url\\(['|\"]*(.*?)['|\"]*\\)");
 
+	private static final String EMBED_BASE64 = "embedBase64";
```
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/resource/CssUrlReplacer.java]: Please make this `public`.
An application developer may want to use ...
And add javadoc what it is about.

File: wicket-core/src/main/java/org/apache/wicket/resource/CssUrlReplacer.java [hunk 2]
```diff
@@ -58,9 +66,12 @@ public String process(String input, Class<?> scope, String name)
 		{
 			Url imageCandidateUrl = Url.parse(matcher.group(1));
 			CharSequence processedUrl;
+			boolean embedded = false;
+
 			if (imageCandidateUrl.isFull())
 			{
 				processedUrl = imageCandidateUrl.toString(Url.StringMode.FULL);
+
```
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/resource/CssUrlReplacer.java]: no need of the new extra line

File: wicket-core/src/main/java/org/apache/wicket/resource/CssUrlReplacer.java [hunk 3]
```diff
@@ -71,21 +82,67 @@ else if (imageCandidateUrl.isContextAbsolute())
...
+						cssUrlCopy.toString().replace("?" + EMBED_BASE64, ""));
+					try
+					{
+						processedUrl = createBase64EncodedImage(imageReference);
+					}
+					catch (Exception e)
+					{
+						throw new WicketRuntimeException(
+							"Error while embedding an image into the css: " +
+								imageReference.toString(), e);
```
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/resource/CssUrlReplacer.java]: no need of `toString()`. It is called implicitly.

File: wicket-core/src/main/java/org/apache/wicket/resource/CssUrlReplacer.java [hunk 4]
```diff
@@ -71,21 +82,67 @@ else if (imageCandidateUrl.isContextAbsolute())
...
+	 * @throws IOException
+	 *             if the stream couldn't be read
+	 */
+	private CharSequence createBase64EncodedImage(PackageResourceReference imageReference)
+		throws ResourceStreamNotFoundException, IOException
+	{
+		IResourceStream resourceStream = imageReference.getResource().getResourceStream();
+		byte[] bytes = new byte[(int)resourceStream.length().bytes()];
+		DataInputStream dataInputStream = new DataInputStream(resourceStream.getInputStream());
+		dataInputStream.readFully(bytes);
```
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/resource/CssUrlReplacer.java]: You can use org.apache.wicket.util.io.IOUtils#toByteArray(java.io.InputStream) instead.

**Comment by martin-g** [comment]
Good job, Tobias!
I've made some minor comments. Please apply the suggested improvements and squash the commits into one.

**Comment by klopfdreh** [comment]
Integrated: https://github.com/apache/wicket/commit/8c9fd4efa2abdd774fd0e396131b4237a3bc528f

**Pull request closed without merging**